docs/proposals/20210222-kubelet-authentication.md

```diff
@@ -0,0 +1,319 @@
+- Kubeadm COULD allow opt-out of kubeadm setting up the bootstrapper and node groups and bindings.
+
+
+#### Kubelet authenticator flow
+
+The authenticator will be responsible for both updating the kubelet client and serving certificates.
+
+![kubelet auth](images/kubelet-authentication/kubelet-auth.png)
+
+
+
+### User Stories
+
+
+
+#### Story 1: machine attestation
+
+A cluster operator has been asked to ensure compliance with NIST SP 800-190 Application Container Security Guide. Hardware countermeasure 4.6 suggests that container platforms should make use of trusted computing. In a Kubernetes context, this would mean providing hardware node attestation wherever possible. Each cloud provider supports different mechanisms for this (see Appendix). (cluster-api/3762
+
+### Requirements (Optional)
+
+The node bootstrapper MUST be able to attest the identity of the machine against a chain of trust provided by the hardware or cloud provider.
+
+
+### Implementation Details/Notes/Constraints
+
+- What are some important details that didn't come across above.
+- What are the caveats to the implementation?
+- Go in to as much detail as necessary here.
+- Talk about core concepts and how they releate.
+
+### Security Model
+
+Document the intended security model for the proposal, including implications
+on the Kubernetes RBAC model. Questions you may want to answer include:
+
+* Does this proposal implement security controls or require the need to do so?
+  * If so, consider describing the different roles and permissions with tables.
+* Are their adequate security warnings where appropriate (see https://adam.shostack.org/ReederEtAl_NEATatMicrosoft.pdf for guidance).
+* Are regex expressions going to be used, and are their appropriate defenses against DOS.
+* Is any sensitive data being stored in a secret, and only exists for as long as necessary?
+
+### Risks and Mitigations
+
+There may be additional security risks being introduced in this design. In order to mitigate this, this proposal will be taken to SIG Security and SIG Auth for review **before the beta graduation**.
+
+## Alternatives
+
+#### Implement within the cloud providers instead of Cluster API
+Given that there is an extant implementation in cloud-provider-gcp, this could be extended to all of the cloud providers. However, there are some advantages to making Cluster API responsible for kubelet registration:
+- Cluster API can create the node resource as part of the CSR approval process and apply the relevant taints, labels and roles up front. The Kubernetes Cloud Provider contract has no understanding of the desired state of a node.
```

Comment:
at one point i was playing with manually creating a Node object and i don't think the kubelet liked that very much.
don't recall all the details, but it never became Ready or something like that.

Reply:
Me and @yastij reviewed the kubelet code, and it seemed to be ok. Nevertheless, we don't need to pre-create the node, so have revised this text.